File: complexity/merge_sort.py
# %%

# runtime complexity: O(n)
# space complexity: O(n)
def merge(left, right):
    # left = [1, 3]
    # right = [2, 5, 7]
    # result = [1, 2, 3,  ]
    result = []
    i = 0
    j = 0
    n = len(left) + len(right)
    while len(result) < n:
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
        if i == len(left):
            result += right[j:]
        if j == len(right):
            result += left[i:]
    return result

# runtime complexity: O(n * log n)
# space complexity: O(log n) + O(n) = O(n)
def merge_sort(arr):
    n = len(arr)
    if n <= 1:
        return arr
    mid = n // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

print(merge_sort([5, 3, 7, 2, 10]))
# %%

File: golang/merge_sort.go
package main

import (
	"fmt"
)

func merge(arr []int, mid int) []int {
	result := []int{}
	var i, j = 0, mid
	n := len(arr)
	for len(result) < n {
		if arr[i] < arr[j] {
			result = append(result, arr[i])
			i += 1
		} else {
			result = append(result, arr[j])
			j += 1
		}
		if i == mid {
			result = append(result, arr[j:]...)
		}
		if j == n {
			result = append(result, arr[i:mid]...)
		}
	}
	return result
}

func merge_sort(arr []int) {

	if len(arr) > 1 {
		mid := len(arr) / 2
		merge_sort(arr[:mid])
		merge_sort(arr[mid:])
		ret := merge(arr, mid)
		for i := range arr {
			arr[i] = ret[i]
		}
	}
}

func printArray(arr []int) {
	for i := range arr {
		fmt.Print(arr[i], ",")
	}
	fmt.Println()
}

func main() {
	arr := []int{5, 3, 7, 2, 10}
	printArray(arr)
	merge_sort(arr)
	fmt.Println("sorted:")
	printArray(arr)
}
